File: experimental/java/src/main/java/com/azure/data/cosmos/serialization/hybridrow/schemas/ObjectPropertyType.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.data.cosmos.serialization.hybridrow.schemas;

import java.util.Collections;
import java.util.List;

/**
 * Object properties represent nested structures.
 * <p>
 * Object properties map to multiple columns depending on the number of internal properties within the defined object
 * structure.  Object properties are provided as a convince in schema design.  They are effectively equivalent to
 * defining the same properties explicitly via {@link PrimitivePropertyType} with nested property paths.
 */
public class ObjectPropertyType extends ScopePropertyType {

    private List<Property> properties;

    /**
     * Initializes a new instance of the {@link ObjectPropertyType} class.
     */
    public ObjectPropertyType() {
        this.properties = Collections.emptyList();
    }

    /**
     * A list of zero or more property definitions that define the columns within the schema.
     *
     * @return a list of zero or more property definitions that define the columns within the schema.
     */
    public final List<Property> properties() {
        return this.properties;
    }

    public final ObjectPropertyType properties(List<Property> value) {
        this.properties = value != null ? value : Collections.emptyList();
        return this;
    }
}

File: src/Serialization/HybridRow/Schemas/ObjectPropertyType.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Schemas
{
    using System.Collections.Generic;
    using Newtonsoft.Json;

    /// <summary>Object properties represent nested structures.</summary>
    /// <remarks>
    /// Object properties map to multiple columns depending on the number of internal properties
    /// within the defined object structure.  Object properties are provided as a convince in schema
    /// design.  They are effectively equivalent to defining the same properties explicitly via
    /// <see cref="PrimitivePropertyType" /> with nested property paths.
    /// </remarks>
    public sealed class ObjectPropertyType : ScopePropertyType
    {
        /// <summary>A list of zero or more property definitions that define the columns within the schema.</summary>
        private List<Property> properties;

        /// <summary>Initializes a new instance of the <see cref="ObjectPropertyType" /> class.</summary>
        public ObjectPropertyType()
            : base(TypeKind.Object)
        {
            this.properties = new List<Property>();
        }

        /// <summary>A list of zero or more property definitions that define the columns within the schema.</summary>
        [JsonProperty(PropertyName = "properties")]
        public List<Property> Properties
        {
            get => this.properties;
            set => this.properties = value ?? new List<Property>();
        }
    }
}
